File: ruby/boyer_moore_string_search/boyer_moore_string_search.rb
def getComparisonIndexes(indexCount, toFind)
  indexes = [-1] * indexCount
  toFind.chars.each_with_index { |c, i|
    indexes[c.ord] = i
  }
  indexes
end

def boyerMooreStringSearch(toSearch, toFind)
  endLen = toFind.size
  maxSearchLen = toSearch.size
  lastfindIndex = endLen - 1
  i = lastfindIndex
  j = lastfindIndex
  indexes = getComparisonIndexes(256, toFind)

  while i < maxSearchLen
    searchVal = toSearch[i]
    findVal = toFind[j]
    if searchVal != findVal
      j = lastfindIndex
      lastIndexInFindOfVal = indexes[searchVal.ord]
      if lastIndexInFindOfVal > -1
        i += endLen - (lastIndexInFindOfVal + 1)
      else
        i += j + 1
      end
    else
      i -= 1
      j -= 1
      return i - j if j == 0
    end
  end

  return -1
end

toFind = "jump"
toSearch = "the quick brown fox jumps over the lazy dog"
puts boyerMooreStringSearch(toSearch, toFind)

toSearch = " jum jumps over the lazy dog"
puts boyerMooreStringSearch(toSearch, toFind)

toSearch = "the quick brown fox over the lazy dog"
puts boyerMooreStringSearch(toSearch, toFind)

toFind = "example"
toSearch = "here is a simple example"
puts boyerMooreStringSearch(toSearch, toFind)

File: javascript/boyer_moore_string_search/boyer_moore_string_search.js
function getComparisonIndexes (indexCount, toFind) {
  let indexes = new Array(indexCount).fill(-1);
  for (let i = 0; i < toFind.length; i++) {
    indexes[ toFind.charCodeAt(i) ] = i;
  }
  return indexes;
}

function boyerMooreStringSearch (toSearch, toFind) {
  let endLen = toFind.length;
  let maxSearchLen = toSearch.length;
  const lastfindIndex = endLen - 1;
  let i = lastfindIndex;
  let j = lastfindIndex;
  let indexes = getComparisonIndexes(256, toFind);

  while (i < maxSearchLen) {
    let searchVal = toSearch[i];
    let findVal = toFind[j];
    if (searchVal != findVal) {
      j = lastfindIndex;
      let lastIndexInFindOfVal = indexes[searchVal.charCodeAt(0)];
      if (lastIndexInFindOfVal > -1) {
        i += endLen - (lastIndexInFindOfVal + 1);
      } else {
        i += j + 1;
      }
    } else {
      i -= 1;
      j -= 1;
      if (j == 0) {
        return i - j;
      }
    }
  }

  return -1;
}

let toFind = "jump";
let toSearch = "the quick brown fox jumps over the lazy dog";
console.log(boyerMooreStringSearch(toSearch, toFind));

toSearch = " jum jumps over the lazy dog";
console.log(boyerMooreStringSearch(toSearch, toFind));

toSearch = "the quick brown fox over the lazy dog";
console.log(boyerMooreStringSearch(toSearch, toFind));

toFind = "example";
toSearch = "here is a simple example";
console.log(boyerMooreStringSearch(toSearch, toFind));